README
# Reporter
![Pub Version](https://img.shields.io/pub/v/reporter)
![Pub Version](https://img.shields.io/pub/v/reporter_html)
![GitHub last commit on main](https://img.shields.io/github/last-commit/Bdaya-Dev/reporter/main)

A standard for generating reports in pure dart.
> Note: Currently only supports tabular reports.

## Features

* Build tabular reports with ease.
* Handles column nesting (colspan).
* Handles row nesting (rowspan).
* Full customizable.
* Language independent, can be adapted for Excel, Html, Pdf, Markdown, txt, etc ...
* Surprisingly easy to use.

## Getting started

1. Depend on the `reporter` package `dart pub add reporter`
2. Optionally depend on other specific generators like `reporter_html`, `reporter_syncfusion_flutter_xlsio`, etc...
3. Use `TabularReporter.calculateCells` to calculate the cells from rows and columns

## Example

Checkout the [example](../reporter_html/example/reporter_html_example.dart) to see how to generate this table:

<table>
    <tr>
        <th rowspan="3">Client</th>
        <th colspan="4">Payments</th>
    </tr>
    <tr>
        <th rowspan='2'>Reason</th>
        <th rowspan='2'>Total To Pay</th>
        <th colspan='2'>Instances</th>
    </tr>
    <tr>
        <th>Date</th>
        <th>Amount</th>
    </tr>
    <tr>
        <td rowspan="6">Person1</td>
        <td rowspan="3">Project1</td>
        <td rowspan="3">5000</td>
        <td>01-08-2023</td>
        <td>2000</td>
    </tr>
    <tr>
        <td>15-08-2023</td>
        <td>1500</td>
    </tr>
    <tr>
        <td>30-08-2023</td>
        <td>1500</td>
    </tr>
    <tr>
        <td>Project2</td>
        <td>1000</td>
        <td></td>
        <td></td>
    </tr>
    <tr>
        <td rowspan="2">Project3</td>
        <td rowspan="2">4000</td>
        <td>01-08-2023</td>
        <td>1000</td>
    </tr>
    <tr>
        <td>15-08-2023</td>
        <td>2000</td>
    </tr>
    <tr>
        <td rowspan="1">Person2</td>
        <td rowspan="1">Project4</td>
        <td rowspan="1">0</td>
    </tr>
</table>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2023-08-20

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.3.0`](#reporter---v030)
 - [`reporter_html` - `v0.3.0`](#reporter_html---v030)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.2.0`](#reporter_syncfusion_flutter_xlsio---v020)

---

#### `reporter` - `v0.3.0`

 - **FIX**: calculation mistakes, and introduce calculateFullTable.

#### `reporter_html` - `v0.3.0`

 - **FEAT**(reporter_html): use calculateFullTable.

#### `reporter_syncfusion_flutter_xlsio` - `v0.2.0`

 - **FEAT**(reporter_syncfusion_flutter_xlsio): use calculateFullTable, and fix range calculations.


## 2023-08-09

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter_syncfusion_flutter_xlsio` - `v0.1.2`](#reporter_syncfusion_flutter_xlsio---v012)
 - [`reporter_html` - `v0.2.1`](#reporter_html---v021)

---

#### `reporter_syncfusion_flutter_xlsio` - `v0.1.2`

 - **FIX**: offset body by 1 row.

#### `reporter_html` - `v0.2.1`

 - **FEAT**: added hooks to html.


## 2023-08-06

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter_html` - `v0.2.0`](#reporter_html---v020)

---

#### `reporter_html` - `v0.2.0`

 - feat: added hooks to html generation config


## 2023-08-06

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter_html` - `v0.1.0`](#reporter_html---v010)

---

#### `reporter_html` - `v0.1.0`

 - change createElementFromValue to accept Node instead of Element

 - **FEAT**: relax createElementFromValue.


## 2023-08-06

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.2.0`](#reporter---v020)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.1.1`](#reporter_syncfusion_flutter_xlsio---v011)
 - [`reporter_html` - `v0.0.6+1`](#reporter_html---v0061)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reporter_html` - `v0.0.6+1`

---

#### `reporter` - `v0.2.0`

 - **FEAT**: add boolean isColumn metadata for cells that represent column headers.

#### `reporter_syncfusion_flutter_xlsio` - `v0.1.1`

 - **FEAT**: support merging excel configs.


## 2023-08-04

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.1.4+1`](#reporter---v0141)
 - [`reporter_html` - `v0.0.6`](#reporter_html---v006)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.1.0+1`](#reporter_syncfusion_flutter_xlsio---v0101)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reporter_syncfusion_flutter_xlsio` - `v0.1.0+1`

---

#### `reporter` - `v0.1.4+1`

 - readme update

#### `reporter_html` - `v0.0.6`

 - **FEAT**: improved tests.


## 2023-08-04

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter_syncfusion_flutter_xlsio` - `v0.1.0`](#reporter_syncfusion_flutter_xlsio---v010)

---

#### `reporter_syncfusion_flutter_xlsio` - `v0.1.0`

 - **FEAT**: add assignFullTableFromRowsAndColumns.


## 2023-08-04

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.1.4`](#reporter---v014)
 - [`reporter_html` - `v0.0.5`](#reporter_html---v005)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.0.2`](#reporter_syncfusion_flutter_xlsio---v002)

---

#### `reporter` - `v0.1.4`

 - fix example link

#### `reporter_html` - `v0.0.5`

 - fix example link

#### `reporter_syncfusion_flutter_xlsio` - `v0.0.2`


## 2023-08-04

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.1.3`](#reporter---v013)
 - [`reporter_html` - `v0.0.4+1`](#reporter_html---v0041)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.0.1+1`](#reporter_syncfusion_flutter_xlsio---v0011)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reporter_html` - `v0.0.4+1`
 - `reporter_syncfusion_flutter_xlsio` - `v0.0.1+1`

---

#### `reporter` - `v0.1.3`

 - readme changes


## 2023-08-04

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`reporter` - `v0.1.2`](#reporter---v012)
 - [`reporter_html` - `v0.0.4`](#reporter_html---v004)
 - [`reporter_syncfusion_flutter_xlsio` - `v0.0.1+1`](#reporter_syncfusion_flutter_xlsio---v0011)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `reporter_syncfusion_flutter_xlsio` - `v0.0.1+1`

---

#### `reporter` - `v0.1.2`

 - Dedicated example project

#### `reporter_html` - `v0.0.4`

 - Dedicated example project


## 2023-08-03


test/models/my_payment_instance.dart
class MyPaymentInstance {
  static const kdate = 'date';
  static const kamount = 'amount';

  final DateTime date;
  final double amount;

  const MyPaymentInstance({
    required this.date,
    required this.amount,
  });
}


lib/src/range.dart
class ReportRange {
  final int rowIndex;
  final int columnIndex;

  int get endRowIndex => rowIndex + rowSpan - 1;
  int get endColumnIndex => columnIndex + colSpan - 1;

  final int rowSpan;
  final int colSpan;

  const ReportRange({
    required this.rowIndex,
    required this.columnIndex,
    required this.rowSpan,
    required this.colSpan,
  });

  const ReportRange.single({
    required this.rowIndex,
    required this.columnIndex,
  })  : rowSpan = 1,
        colSpan = 1;

  const ReportRange.fromIndexes({
    required int startRowIndex,
    required int startColumnIndex,
    required int endRowIndex,
    required int endColumnIndex,
  })  : rowSpan = endRowIndex - startRowIndex + 1,
        colSpan = endColumnIndex - startColumnIndex + 1,
        rowIndex = startRowIndex,
        columnIndex = startColumnIndex;

  ReportRange copyWith({
    int? rowIndex,
    int? columnIndex,
    int? rowSpan,
    int? colSpan,
  }) {
    return ReportRange(
      rowIndex: rowIndex ?? this.rowIndex,
      columnIndex: columnIndex ?? this.columnIndex,
      rowSpan: rowSpan ?? this.rowSpan,
      colSpan: colSpan ?? this.colSpan,
    );
  }

  ReportRange operator +(ReportRange other) {
    return ReportRange(
      rowIndex: rowIndex + other.rowIndex,
      columnIndex: columnIndex + other.columnIndex,
      rowSpan: rowSpan + other.rowSpan,
      colSpan: colSpan + other.colSpan,
    );
  }

  ReportRange shiftRowsBy(int rows) {
    return copyWith(
      rowIndex: rowIndex + rows,
    );
  }
  ReportRange shiftColumnsBy(int columns) {
    return copyWith(
      columnIndex: columnIndex + columns,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is ReportRange &&
        other.rowIndex == rowIndex &&
        other.columnIndex == columnIndex &&
        other.rowSpan == rowSpan &&
        other.colSpan == colSpan;
  }

  @override
  int get hashCode {
    return rowIndex.hashCode ^
        columnIndex.hashCode ^
        rowSpan.hashCode ^
        colSpan.hashCode;
  }

  @override
  String toString() {
    return '[$rowIndex..$endRowIndex][$columnIndex..$endColumnIndex]';
  }
}


lib/src/column.dart
import 'package:collection/collection.dart';

/// Represents a column in a tabular report.
/// A column can contain nested columns, which corresponds to a change in [columnSpan].
class ReportColumn {
  final String id;
  final String name;
  final List<ReportColumn> children;
  final Map<String, dynamic> metadata;

  int get columnSpan =>
      children.isEmpty ? 1 : children.map((e) => e.columnSpan).sum;

  int get maxDepth =>
      (children.isEmpty ? 0 : children.map((e) => e.maxDepth).max) + 1;

  /// A convenience constructor when the id is the same as the name, and they are both unique.
  const ReportColumn.nameOnly({
    required this.name,
    this.children = const [],
    this.metadata = const {},
  }) : id = name;

  const ReportColumn({
    required this.id,
    required this.name,
    this.children = const [],
    this.metadata = const {},
  });

  static Iterable<ReportColumn> _expand(ReportColumn column) sync* {
    yield column;
    yield* column.children.map(_expand).flattened;
  }

  static Map<String, ReportColumn> generateColumnMap(
    Iterable<ReportColumn> columns,
  ) {
    return Map.fromEntries(
      columns.expand(_expand).map((e) => MapEntry(e.id, e)),
    );
  }

  static Iterable<ReportColumn> getColumnsWithNoChildren(
    Iterable<ReportColumn> columns,
  ) {
    return columns.expand(
      (e) => e.children.isEmpty ? [e] : getColumnsWithNoChildren(e.children),
    );
  }

  static Map<String, Set<String>> generateParentsMap(
    Iterable<ReportColumn> columns,
  ) {
    final res = <String, Set<String>>{};
    for (var c in columns) {
      _fillParentsMap(c, res);
    }
    return res;
  }

  static void _fillParentsMap(
    ReportColumn parent,
    Map<String, Set<String>> res,
  ) {
    final parentParents = res[parent.id] ??= {};
    for (var child in parent.children) {
      final parents = res[child.id] ??= {};
      parents.addAll(parentParents);
      parents.add(parent.id);
      _fillParentsMap(child, res);
    }
  }
}

// class TypedReportColumn<T> extends ReportColumn {
//   const TypedReportColumn.fromAssignment({
//     required super.id,
//     required super.name,
//     required this.getAssigment,
//     super.children,
//     super.metadata,
//   });

//   TypedReportColumn({
//     required super.id,
//     required super.name,
//     required dynamic Function(T item) getValue,
//     super.children,
//     super.metadata,
//   }) : getAssigment = ((T item) =>
//             ReportColumnValueAssignment(value: getValue(item)));

//   final ReportColumnValueAssignment? Function(T item) getAssigment;
// }


lib/src/calculated_range.dart
import 'range.dart';

/// Represents a range with useful information, like the value and the metadata.
class ReportCalculatedRange {
  /// The range
  final ReportRange range;

  /// All the columns this cell belongs to
  final Set<String> columnIds;

  /// The column that's directly above this cell
  final String directColumnId;

  final Object? value;
  final Map<String, dynamic> metadata;

  const ReportCalculatedRange({
    required this.range,
    required this.columnIds,
    required this.directColumnId,
    required this.value,
    this.metadata = const {},
  });

  ReportCalculatedRange copyWith({
    ReportRange? range,
    Set<String>? columnIds,
    String? directColumnId,
    Object? value,
    Map<String, dynamic>? metadata,
  }) {
    return ReportCalculatedRange(
      range: range ?? this.range,
      columnIds: columnIds ?? this.columnIds,
      directColumnId: directColumnId ?? this.directColumnId,
      value: value ?? this.value,
      metadata: metadata ?? this.metadata,
    );
  }

  ReportCalculatedRange copyWithValue(Object? value) {
    return ReportCalculatedRange(
      range: range,
      columnIds: columnIds,
      directColumnId: directColumnId,
      value: value,
      metadata: metadata,
    );
  }

  @override
  String toString() {
    return '$range = $value ~ $metadata';
  }
}


lib/src/tabular_reporter.dart
import 'package:reporter/reporter.dart';

/// A utility class for generating tabular reports (like spreadsheets)
/// from arbitrary data.
class TabularReporter {
  /// Takes the rows and columns and transforms them to ranges with values.
  static List<ReportCalculatedRange> calculateCells({
    required List<ReportRow> rows,
    required List<ReportColumn> columns,
    int offsetRowIndex = 0,
    int offsetColumnIndex = 0,
  }) {
    if (rows.isEmpty) {
      return [];
    }
    final leafColumns = ReportColumn.getColumnsWithNoChildren(columns).toList();
    final reverseColumnIndex =
        leafColumns.asMap().map((key, value) => MapEntry(value.id, key));
    final leafMap = Map.fromEntries(leafColumns.map((e) => MapEntry(e.id, e)));
    final parentsMap = ReportColumn.generateParentsMap(columns);
    //map column to its parents

    final res = <ReportCalculatedRange>[];
    _calculateCells(
      leafMap: leafMap,
      parentsMap: parentsMap,
      res: res,
      reverseColumnIndex: reverseColumnIndex,
      rows: rows,
      offsetColumnIndex: offsetColumnIndex,
      offsetRowIndex: offsetRowIndex,
    );
    return res;
  }

  static void _calculateCells({
    required Map<String, ReportColumn> leafMap,
    required List<ReportRow> rows,
    required List<ReportCalculatedRange> res,
    required Map<String, int> reverseColumnIndex,
    required Map<String, Set<String>> parentsMap,
    required int offsetRowIndex,
    required int offsetColumnIndex,
  }) {
    int visibleRowIndex = offsetRowIndex;
    for (var rowIndex = 0; rowIndex < rows.length; rowIndex++) {
      final row = rows[rowIndex];
      for (var assignmentEntry in row.columnAssignments.entries) {
        final columnId = assignmentEntry.key;
        final column = leafMap[columnId];
        final columnIndex = reverseColumnIndex[columnId];
        if (column == null || columnIndex == null) {
          continue;
        }

        res.add(
          ReportCalculatedRange(
            directColumnId: columnId,
            metadata: assignmentEntry.value.metadata,
            value: assignmentEntry.value.value,
            columnIds: {
              ...?parentsMap[columnId],
              columnId,
            },
            range: ReportRange(
              rowIndex: visibleRowIndex,
              columnIndex: offsetColumnIndex + columnIndex,
              rowSpan: row.rowSpan,
              colSpan: column.columnSpan,
            ),
          ),
        );
      }
      //repeat the process for child rows
      _calculateCells(
        rows: row.children,
        leafMap: leafMap,
        parentsMap: parentsMap,
        res: res,
        reverseColumnIndex: reverseColumnIndex,
        offsetColumnIndex: offsetColumnIndex,
        offsetRowIndex: visibleRowIndex,
      );

      visibleRowIndex += row.rowSpan;
    }
  }

  /// makes sure all the ranges are single, for cases where Excel does't want the data merged.
  /// Note that this keeps the metadata the same for all the split cells.
  List<ReportCalculatedRange> enforceSingleCell(
    Iterable<ReportCalculatedRange> ranges, {
    bool splitCellsAreEmpty = false,
  }) {
    final res = <ReportCalculatedRange>[];
    for (var element in ranges) {
      final range = element.range;
      if (range.rowSpan <= 1 && range.colSpan <= 1) {
        res.add(element);
        continue;
      }
      //split
      for (var r = 0; r < range.rowSpan; r++) {
        for (var c = 0; c < range.colSpan; c++) {
          res.add(
            element
                .copyWith(
                  range: ReportRange.single(
                    rowIndex: range.rowIndex + r,
                    columnIndex: range.columnIndex + c,
                  ),
                )
                .copyWithValue(
                  splitCellsAreEmpty && (r > 0 || c > 0) ? null : element.value,
                ),
          );
        }
      }
    }
    return res;
  }
}


example/reporter_example.dart
import 'package:reporter/reporter.dart';

/// In this example we use a client that has 0 or more payment groups,
/// where each group has 0 or more payment instances.
class MyClient {
  static const kname = 'name';

  final String name;
  final List<MyPaymentGroup> paymentGroups;

  const MyClient({
    required this.name,
    this.paymentGroups = const [],
  });
}

class MyPaymentGroup {
  static const kreason = 'reason';
  static const ktotalToPay = 'totalToPay';

  final String reason;
  final double totalToPay;
  final List<MyPaymentInstance> paymentInstances;

  const MyPaymentGroup({
    required this.reason,
    required this.totalToPay,
    this.paymentInstances = const [],
  });
}

class MyPaymentInstance {
  static const kdate = 'date';
  static const kamount = 'amount';

  final DateTime date;
  final double amount;

  const MyPaymentInstance({
    required this.date,
    required this.amount,
  });
}

/// Define the structure of the columns, note that only the columns with no children
/// will correspond to actual data, this their [id] is enforced.
final columns = <ReportColumn>[
  ReportColumn(
    id: MyClient.kname,
    name: 'Client',
  ),
  ReportColumn.nameOnly(
    name: 'payments',
    children: [
      ReportColumn(
        id: MyPaymentGroup.kreason,
        name: 'reason',
      ),
      ReportColumn(
        id: MyPaymentGroup.ktotalToPay,
        name: 'total to pay',
      ),
      ReportColumn.nameOnly(
        name: 'instances',
        children: [
          ReportColumn(
            id: MyPaymentInstance.kdate,
            name: 'date',
          ),
          ReportColumn(
            id: MyPaymentInstance.kamount,
            name: 'amount',
          ),
        ],
      ),
    ],
  ),
];

/// This maps the user data to [ReportRow] which contains grouping information about
/// the rows.
List<ReportRow> mapDataToRows(
  List<MyClient> clients,
) {
  return clients.map((client) {
    return ReportRow(
      // Metadata contains an arbitrary map, which might be useful later when generating
      // Excel/Html tables for example.
      metadata: {
        'color': 'green',
        'src': client,
      },
      // columnAssignments are the data that correspond to each column.
      columnAssignments: {
        MyClient.kname: ReportColumnValueAssignment(value: client.name),
      },
      // children represent the nested rows.
      // There is NO limit on the degree of nesting.
      children: client.paymentGroups.map((paymentGroup) {
        return ReportRow(
          columnAssignments: {
            MyPaymentGroup.kreason: ReportColumnValueAssignment(
              value: paymentGroup.reason,
            ),
            MyPaymentGroup.ktotalToPay:
                ReportColumnValueAssignment(value: paymentGroup.totalToPay),
          },
          children: paymentGroup.paymentInstances.map((paymentInstance) {
            return ReportRow(
              columnAssignments: {
                MyPaymentInstance.kamount:
                    ReportColumnValueAssignment(value: paymentInstance.amount),
                MyPaymentInstance.kdate:
                    ReportColumnValueAssignment(value: paymentInstance.date),
              },
            );
          }).toList(),
        );
      }).toList(),
    );
  }).toList();
}

void main() {
  //The data source, which usually comes from the server.
  final sampleData = <MyClient>[
    MyClient(
      name: 'person1',
      paymentGroups: [
        MyPaymentGroup(
          reason: 'project1',
          totalToPay: 5000,
          paymentInstances: [
            MyPaymentInstance(amount: 2000, date: DateTime(2023, 08, 01)),
            MyPaymentInstance(amount: 1500, date: DateTime(2023, 08, 15)),
            MyPaymentInstance(amount: 1500, date: DateTime(2023, 08, 30)),
          ],
        ),
        MyPaymentGroup(
          reason: 'project2',
          totalToPay: 1000,
        ),
        MyPaymentGroup(
          reason: 'project3',
          totalToPay: 4000,
          paymentInstances: [
            MyPaymentInstance(amount: 1000, date: DateTime(2023, 09, 1)),
            MyPaymentInstance(amount: 2000, date: DateTime(2023, 09, 15)),
          ],
        ),
      ],
    ),
    MyClient(
      name: 'person2',
      paymentGroups: [
        MyPaymentGroup(reason: 'project4', totalToPay: 0),
      ],
    ),
  ];

  // calculates the cells based on the input rows and columns.
  // These cells can be adapted to different generators,
  // e.g.: excel, pdf, markdown, html, etc...
  final List<ReportCalculatedRange> cells = TabularReporter.calculateCells(
    columns: columns,
    rows: mapDataToRows(sampleData),
    //(default=0) Shift all columns by 1, useful for inserting columns before the table
    offsetColumnIndex: 1,
    //(default=0) Shift all rows by 2, useful for inserting rows before the table
    offsetRowIndex: 2,
  );

  print('Generated Cells: ${cells.length}');
}
